### PR TITLE
fix: replace node buffers with uint8arrays

### PR DESCRIPTION
**Pull request by achingbrain**
Relaxes input from requiring node `Buffer`s to being `Uint8Arrays`.

This also means that the `.buffer` (now `.byte`) and `.prefix` properties have changed to be `Uint8Array`s.

Depends on:

- [x] https://github.com/multiformats/js-multicodec/pull/63
- [x] https://github.com/multiformats/js-multihashing-async/pull/78
- [x] https://github.com/multiformats/js-multihash/pull/92

BREAKING CHANGES:

- node `Buffer`s have been replaced with `Uint8Array`s
- the `.buffer` property has been renamed to `.bytes` and is now a `Uint8Array`
- the `.prefix` property is now a `Uint8Array`

### CODE REVIEW

**Comment by vmx** [comment]
The CI failure is real, though you could also use https://github.com/Gozala/web-encoding/ instead of ipfs-utils.

**Comment by achingbrain** [comment]
> you could also use Gozala/web-encoding instead of ipfs-utils.

I will if using `ipfs-utils` would prevent this from getting merged, but `ipfs-utils` already has the TextEncoder browser/node polyfill in it so it seems redundant.

File: test/index.spec.js [hunk 1]
```diff
@@ -2,21 +2,35 @@
...
+  }
+
+  const parts = []
+
+  for (let i = 0; i < str.length; i += 2) {
+    parts.push(parseInt(`${str[i]}${str[i + 1]}`, 16))
+  }
+
+  return Uint8Array.from(parts)
+}
```
Comment (Gozala) [inline test/index.spec.js:27]: 📝 I remember having to do this kind of thing elsewhere, maybe it's worth extracting it into HexEncoder / HexDecoder somewhere for sharing.

**Comment by Gozala** [comment]
> I will if using `ipfs-utils` would prevent this from getting merged, but `ipfs-utils` already has the TextEncoder browser/node polyfill in it so it seems redundant.
> 
> Plus once [ipfs/js-ipfs-utils#55](https://github.com/ipfs/js-ipfs-utils/pull/55) is merged I can remove some of the duplicated utility functions here.

For what it's worth we have discussed option of using `ipfs-utils` vs creating a `web-encodings` library and chose later to avoid pulling all the extra stuff that `api-utils` has. I'm fine with either option.

File: README.md [hunk 2]
```diff
@@ -142,19 +158,19 @@ Property containing the CID version integer.
...
 #### cid.multihash
 
-Property containing the multihash buffer.
+Property containing the multihash `Uint8Array`.
 
 #### cid.multibaseName
 
 Property containing the default base to use when calling `.toString`
 
 #### cid.buffer
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply README.md]: It makes me wince a little, because it's going to be very disruptive, but I've renamed it to `.bytes`.

File: src/index.js [hunk 3]
```diff
@@ -155,11 +154,11 @@ class CID {
       if (this.version === 0) {
         buffer = this.multihash
       } else if (this.version === 1) {
-        buffer = Buffer.concat([
-          Buffer.from('01', 'hex'),
-          multicodec.getCodeVarint(this.codec),
-          this.multihash
-        ])
+        const codec = multicodec.getCodeVarint(this.codec)
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply src/index.js:158]: Done!

File: src/index.js [hunk 4]
```diff
@@ -174,15 +173,19 @@ class CID {
...
   get prefix () {
-    return Buffer.concat([
-      Buffer.from(`0${this.version}`, 'hex'),
-      multicodec.getCodeVarint(this.codec),
-      mh.prefix(this.multihash)
-    ])
+    const codec = multicodec.getCodeVarint(this.codec)
+    const multihash = mh.prefix(this.multihash)
+
+    const prefix = new Uint8Array(codec.byteLength + multihash.byteLength + 1)
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply src/index.js]: Done!

File: test/index.spec.js (hunk 1 above)
Comment (achingbrain) [reply test/index.spec.js]: Done!  Well, sort of - the `toString` and `fromString` functions take `base16` as an encoding.

**Comment by achingbrain** [comment]
I've pulled out `ipfs-utils` in favour of `uint8arrays` which uses `web-encoding` internally.

**Comment by vmx** [comment]
@mikeal @rvagg I'd really like to have your approval on that change. Once merged I want to release this as 1.0.0.

**Comment by mikeal** [comment]
So, +1 in general, great to see this happening.

One concern I have taking a breaking change is just wasting the opportunity to do a bit more. Knowing that we have a more substantial CID change coming in the future it would be great to get some code into this change that will make that transition a bit easier.

For instance, we could add the `.code` property to `CID` and we could allow for instantiation using integers for the codec rather than just strings. This would allow people to migrate to using these newer forms even with the old API and should make the future transition to `js-multiformats` a bit easier.

Basically, we know that we need to move to using integers rather than strings for codec identifiers, it would be nice if we had a transitional release that accepted both so people could upgrade gradually or at least have the opportunity to do some of that migration when they migrate off of Buffer.

**Comment by achingbrain** [comment]
> One concern I have taking a breaking change is just wasting the opportunity to do a bit more

I can see the temptation but I really think to deliver this change in a reasonable timeframe we should limit the scope to what's currently in this PR.

**Comment by Gozala** [comment]
For what it's worth I would also like to slot in `asCID` into 1.0. That said, I do not think we should block this pull request on other things we want in 1.0. We should instead consider having a meeting to make some decision about how to best roll out all the breaking changes.
